Rescu itscentral pipeline

gabrielwol opened the pull request:
## What this pull request accomplishes:

A summary of improvements over the old `rescu` schema:  

**1. Data Availability**  
- The data pipeline is now pulling from farther upstream which has increased data availability. 
    Daily row counts from the two tables over a 4 month period comparing `rescu.volumes_15min` and `vds.counts_15min` for `division_id = 2`:
![rescu vs vds count](https://github.com/CityofToronto/bdit_data-sources/assets/80077912/1b11e5fb-19c2-4527-96ba-34c604618bc9)


**2. Raw data Availability**
- 20 seconds data for RESCU is now included via the `vds.raw_vdsdata` table which allows investigation into the accuracy of 15 minute counts.
- The table `vds.counts_15min` (replaces `rescu.volumes_15min`) now includes extra columns derived from `vds.raw_vdsdata` (num_lanes, expected_bins, num_obs, num_distinct_lanes) to enable accuracy checks. 
- The table `vds.counts_15min_bylane` includes the 15 minute detectors counts by lane, which can be used to verify that all lane sensors are working, or to investigate lane by lane traffic distribution.  

**3. New dataset: `vdsvehicledata`**
- Rescu detectors record individual vehicle detection events under `vds.raw_vdsvehicledata` which allows us to investigate the distribution of vehicle speeds and lengths (and possibly following distance?).  

**4. Detector Inventory Revived**  
- The mysterious `rescu.detector_inventory` table (archived as `vds.detector_inventory_cursed`) is replaced by live configuration tables being pulled from ITS Central: `vds.vdsconfig` for detector details, `vds.entity_locations` for detector locations and new manual view: `vds.detector_inventory` for easy filtering. 

**5. Additional Detector Classes**
- The pipeline now includes many additional vehicle detection systems (VDS) which were not available in the previous rescu schema including: Blue City, SmartMicro and Traffic Signal detectors, explained more under heading [Overview of Sensor Classes](https://github.com/CityofToronto/bdit_data-sources/blob/cdd0c2aacf6f40373da485dc1e9bd60841167b96/volumes/vds/readme.md#overview-of-sensor-classes).  

## Issue(s) this solves:
- Closes #603 
- Closes #638 
- Closes #660 
- Closes #698

## What, in particular, needs to reviewed:
- My first DAG, review everything! 
-
## What needs to be done by a sysadmin after this PR is merged
- rename `rescu` to `rescu_cursed`
_E.g.: these tables need to be migrated/created in the production schema._

tahaislam commented on dags/vds_pull.py:
```diff
@@ -0,0 +1,131 @@
+vds_bot = PostgresHook("vds_bot")
+
+try:
+    repo_path = '/home/gwolofs/bdit_data-sources'
```
This shouldn't be hard-coded or pointing to personal home folders. Check this code snippet as an example of a better implementation https://github.com/Toronto-Big-Data-Innovation-Team/vfh_data_mgmt/blob/bab9e43bda42215b6193f2b710f8cade1eb37e06/dags/unit_validation.py#L23

tahaislam commented on dags/vds_pull.py:
```diff
@@ -0,0 +1,131 @@
+    'email_on_retry': False,
+    'retries': 1,
+    'retry_delay': timedelta(minutes=60),
+    #'on_failure_callback': task_fail_slack_alert,
```
Should we uncomment this?

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,352 @@
+    try: 
+        with itsc_conn.get_conn() as con:
+            raw_data = pd.read_sql(raw_sql, con)
+            LOGGER.info('Fetching vdsdata')
```
move before `pd.read_sql` to let people know that it's reading the data (in case that takes some time to execute)

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,352 @@
+                                        divisionid, vdsid, datetime_20sec, datetime_15min, lane, speedKmh, volumeVehiclesPerHour, occupancyPercent
+                                        ) VALUES %s;''')
+                execute_values(cur, insert_query, data_tuples)
+                LOGGER.info('Inserting vdsdata into RDS.')
```
same as above. move this before the insertion

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,352 @@
+            with con.cursor() as cur:
+                # Drop records for the current date
+                #drop_query = sql.SQL(f"DELETE FROM vds.raw_vdsdata WHERE datetime_bin::date = '{dt}'")
+                #con.execute(drop_query)
```
Why is this query commented?

tahaislam replied:
It depends on how the DAG works. Some DAGs might be updating the entire dataset every time they run. Hence, no need to delete old/incomplete data. Some other DAGs require us to delete this data manually before backfilling (like miovision). But I think your approach here is more robust.

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,352 @@
+#contains individual vehicle activations from highway sensors (speed, length)
+   
+    from psycopg2 import sql, Error
+    from psycopg2.extras import execute_values
```
all imports go to the top of the file

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,352 @@
+        AND TIMEZONE('UTC', d.timestamputc) < {start}::timestamptz + INTERVAL '1 DAY'
+        AND substring(c.sourceid, 1, 3) <> 'BCT'; --bluecity.ai sensors have no data
+    ''').format(
+        start = sql.Literal(start_date + ' 00:00:00 EST5EDT')
```
Please, use double quotes

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,352 @@
+            lane = mv[i][0]
+
+            # Get speed
+            #Stored in km/h * 100. 65535 for null value. Convert 0 to null to maintain backward compatibility
```
Do you mean that the multiplier `100. 65535` is only applied on null values? Or is it applied to all records?

tahaislam commented on volumes/vds/sql/create-function-aggregate_15min_vds_volumes.sql:
```diff
@@ -0,0 +1,43 @@
+        d.vdsid,
+        c.detector_id,
+        d.datetime_15min,
+        SUM(d.volumevehiclesperhour) / 4 / 45 AS volume_15min
```
Can you explain why you need this 45?

tahaislam commented on volumes/vds/sql/create-function-aggregate_15min_vds_volumes.sql:
```diff
@@ -0,0 +1,43 @@
+        d.divisionid,
+        d.vdsid,
+        c.detector_id,
+        d.datetime_15min
```
Did you already validate that each hour has 4 15-min hourly volume? Or should you add a condition with `HAVING COUNT(*) = 4` for instance?

tahaislam commented on volumes/vds/sql/create-table-vdsdata.sql:
```diff
@@ -0,0 +1,16 @@
+    datetime_20sec timestamp without time zone,
+    datetime_15min timestamp without time zone,
+    lane integer, 
+    speedKmh float, 
```
avoid using uppercase letters. It's better to split words with `_`

tahaislam commented on volumes/vds/sql/create-table-vdsdata.sql:
```diff
@@ -0,0 +1,16 @@
+    datetime_15min timestamp without time zone,
+    lane integer, 
+    speedKmh float, 
+    volumeVehiclesPerHour integer,
```
As above. Also the next two lines

tahaislam commented on volumes/vds/sql/create-table-vdsdata_div8001.sql:
```diff
@@ -0,0 +1,35 @@
+    datetime_15min timestamp without time zone,
+    lane integer, 
+    speedKmh float, 
+    volumeVehiclesPerHour integer,
```
Don't use uppercase letters

tahaislam commented on dags/vds_pull.py:
```diff
@@ -0,0 +1,131 @@
+# Get slack member ids
+#dag_owners = Variable.get('dag_owners', deserialize_json=True)
+#names = dag_owners.get(dag_name, ['Unknown']) #find dag owners w/default = Unknown    
+names = ['gabe']
```
Don't forget to update this before the final merge. I updated the Airflow `dag_owners` variable with `"vds_pull": ["gabe"]`

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,352 @@
+        timestamputc >= extract(epoch from timestamp with time zone {start})
+        AND timestamputc < extract(epoch from timestamp with time zone {start} + INTERVAL '1 DAY')
+        AND d.divisionid = 2; --other is 8001 which are traffic signal detectors and are mostly empty.
+    ''').format(
```
Please, use double quotes

tahaislam reviewed (changes requested):
That looks great. I left some comments for you to check. Some of these comments apply to some other locations in similar files, like avoiding using uppercase letters in naming columns.
Also, it'd be helpful if you can add a README to describe the purpose of this pipeline and what each task does. That will help me reviewing the DAG and the `vds_function` file.
There are some sqlfluff comments that look relevant and needed to be addressed.

gabrielwol commented:
May need to add consideration for data backfilled into ITSC.

gabrielwol commented:
@tahaislam 113 commits later, ready for re-review! Some major changes: broke up into multiple dags, added new summary tables, new columns to existing volumes_15 table, speed improvements through query design, added late arriving data monitoring, batch inserting.

tahaislam commented on volumes/vds/readme.md:
```diff
@@ -0,0 +1,255 @@
+# Vehicle Detector System (VDS) data  
+
+VDS data is pulled daily from ITS Central database on terminal server by dags running on Morbius.  
```
I think the ITSC database is accessible through the terminal server or any other on-prem server but not on the terminal server.

tahaislam commented on volumes/vds/readme.md:
```diff
@@ -0,0 +1,255 @@
+# Vehicle Detector System (VDS) data  
+
```
If possible, can you add a table of contents to this page!

tahaislam commented on volumes/vds/readme.md:
```diff
@@ -0,0 +1,255 @@
+# Table Structure  
+## vds.raw_vdsdata
+This table contains parsed data from ITSC public.vdsdata. 
+Volumes are in vehicles per hour for the 20 sec bin. To convert to 15 minute volume, group by datetime_15min and take `SUM(volume_veh_per_hr) / 4 / 45` where / 4 represents hourly to 15 min conversion and / 45 represents number of 20 sec bins in a 15 minute period. This assumes both missing bins and zero values are zeros, in line with old pipeline. 
```
I think the word volume is more common for hourly volumes and it might be better to refer to the 15-min data as 15-minute counts not volumes to avoid confusion.

tahaislam commented on volumes/vds/readme.md:
```diff
@@ -0,0 +1,255 @@
+
+## vds.volumes_15min_bylane
+Same as above but by lane. 
+Will be used to determine when individual lane sensors are down. 
```
Does lane numbering start from left or right?

tahaislam commented on volumes/vds/readme.md:
```diff
@@ -0,0 +1,255 @@
+
+# Dag Design 
+
+## vds_pull_vdsdata [../../../dags/vds_pull_vdsdata.py]
```
If you want to link the text to another file, I think you should have it like this: `[text](link)`, e.g. in this example: `[vds_pull_vdsdata](../../../dags/vds_pull_vdsdata.py)`. But how would this look like as a heading 🤔

tahaislam commented on dags/vds_pull_vdsdata.py:
```diff
@@ -0,0 +1,153 @@
+                    datetime_15min >= '{{ds}} 00:00:00'::timestamp
+                    AND datetime_15min < '{{ds}} 00:00:00'::timestamp + INTERVAL '1 DAY'""",
+            task_id='delete_vdsdata',
+            dag=dag,
```
no need to pass the `dag` to these tasks as long as they are already defined within the DAG context

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,464 @@
+    #pull data in batches and transform + insert.
+    try:
+        with itsc_conn.get_conn() as con:
+            with con.cursor() as cur:
```
you can combine these `with`s to reduce the nesting levels like this:
`with itsc_conn.get_conn() as con, con.cursor() as cur:`

tahaislam commented on volumes/vds/sql/functions/create-function-aggregate_15min_vds_volumes.sql:
```diff
@@ -0,0 +1,64 @@
+                THEN 3 --5 min bins
+            END AS expected_bins
+        FROM vds.vdsconfig 
+    )
```
Can you document this part in a block comment at the top of the function?

tahaislam commented on volumes/vds/sql/functions/create-function-aggregate_15min_vds_volumes.sql:
```diff
@@ -0,0 +1,64 @@
+        c.detector_id,
+        c.lanes AS num_lanes,
+        d.datetime_15min,
+        SUM(d.volume_veh_per_hr) / 4 / c.expected_bins AS volume_15min,
```
can this be renamed as count_15min instead of volume?

tahaislam commented on volumes/vds/sql/functions/create-function-aggregate_15min_vds_volumes_bylane.sql:
```diff
@@ -0,0 +1,63 @@
+            -- / 4 to convert hourly volume to 15 minute volume
+            -- / (expected_bins) to get average 15 minute volume depending on bin size (assumes blanks are 0)
+        c.expected_bins,
+        COUNT(*) AS num_obs
```
so we can get an indication of the quality of these averages based on how much expected_bins and num_obs are close, right? Don't remember if I saw this in the README. should we highlight that some of these 15-min counts might not be so accurate?

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,464 @@
+    except Error as exc:
+        LOGGER.critical(f"Error fetching from {table_name}.")
+        LOGGER.critical(exc)
+        con.close()
```
`with` should close the connection if there's any error. I don't think you need to close it here

tahaislam commented on dags/vds_pull_vdsvehicledata.py:
```diff
@@ -0,0 +1,132 @@
+# Get DAG Owner
+#dag_owners = Variable.get('dag_owners', deserialize_json=True)
+#names = dag_owners.get(dag_name, ['Unknown']) #find dag owners w/default = Unknown    
+names = ['gabe']
```
I updated the `dag_owners` variable on morbius if you want to start reading it

tahaislam commented on dags/vds_pull_vdsvehicledata.py:
```diff
@@ -0,0 +1,132 @@
+                    datetime_15min >= '{{ds}} 00:00:00'::timestamp
+                    AND datetime_15min < '{{ds}} 00:00:00'::timestamp + INTERVAL '1 DAY'""",
+            task_id='delete_veh_speed_data',
+            dag=dag,
```
no need to pass the `dag` to each task

tahaislam commented on volumes/vds/sql/functions/create-function-aggregate_15min_veh_lengths.sql:
```diff
@@ -0,0 +1,39 @@
+
+BEGIN
+	
+    --Aggregated into speed bins and 1 hour bin
```
not speed bins

tahaislam commented on volumes/vds/exploration/time_gaps.sql:
```diff
@@ -0,0 +1,30 @@
+--sensors in the same vdsdata table have different time gaps, ie. some 20sec, some 5 min. 
+--No attribute present in 
```
Do the numbers make sense for the detectors with 5-minute gaps?

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,464 @@
+        webhook_token=slack_webhook_token,
+        message=slack_msg,
+        username='airflow',
+        proxy='http://'+BaseHook.get_connection('slack').password+'@137.15.73.132:8080',
```
I updated the VFH code and Airflow `slack` connection to avoid hard-coding the proxy IP address. Check it [here](https://github.com/Toronto-Big-Data-Innovation-Team/vfh_data_mgmt/blob/133d989ba882e0d3844aa98988b15e7ab5757663/utils/dag_functions.py#L171-L174) if you want to apply it to this function

tahaislam commented on volumes/vds/py/vds_functions.py:
```diff
@@ -0,0 +1,464 @@
+                data = cur.fetchall()
+                LOGGER.info(f"Number of rows fetched from {table_name} table: {len(data)}")
+                data = pd.DataFrame(data)
+                data.columns=[x.name for x in cur.description]
```
why not using `pandas.read_sql`? Check its details [here](https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html)

tahaislam reviewed (changes requested):
Very good work. I left some comments for you to check. Can you also fix sqlfluff warnings?

Tip: `sqlfluff` is installed in `airflow_venv` and you can download its configurations from [here](https://github.com/Toronto-Big-Data-Innovation-Team/.github/blob/main/.sqlfluff), if you want to run `sqlfluff fix`

gabrielwol commented:
@tahaislam can you please rereview - mainly to check responses to previous comments? One of the other major changes I made was moving longer sqls into files for a more concise dag/py file. I also added some more logging of row counts at various points. 
sqlfluff: there are 25 "unparsable" errors which I can't figure out. Nate also unsure. I think we can ignore.

tahaislam commented on volumes/vds/sql/delete/delete-counts_15min.sql:
```diff
@@ -0,0 +1,4 @@
+DELETE FROM vds.counts_15min
+WHERE
+    datetime_15min >= '{{ ds }} 00:00:00'::timestamp
```
You can hide these errors (as long as we know they are actually not errors) by adding `-- noqa: TMP` at the end of the line OR `-- noqa: disable=TMP` at the top of the file

tahaislam commented on volumes/vds/sql/functions/create-function-partition_vdsdata.sql:
```diff
@@ -0,0 +1,104 @@
+Return: Void
+Purpose:
+Create new partitions for the given division_id and year (if needed) under the parent table `base_table`
+*/
```
Thanks for adding this Sphinx-friendly block comment. You can find more about these comments at https://pypi.org/project/sphinx-sql/

tahaislam commented on volumes/vds/sql/functions/create-function-partition_vdsdata.sql:
```diff
@@ -0,0 +1,104 @@
+IF div_id = 8001 THEN --create two partitions instead of 1
+
+    part_table_1 = base_table || '_' || year_::text || '_01_06';
+    part_table_2 = base_table || '_' || year_::text || '_07_12';
```
Why `_01_06` and `_07_12`? Why not `_01_01` and `_07_01`?

tahaislam commented on dags/vds_pull_vdsvehicledata.py:
```diff
@@ -21,13 +21,20 @@
 
+repo_path = os.path.abspath(os.path.dirname(os.path.dirname(os.path.realpath(__file__))))
+sys.path.insert(0, repo_path)
+sys.path.insert(0,os.path.join(repo_path,'volumes/vds/py'))
```
Same comment as in the previous file

tahaislam commented on dags/vds_pull_vdsvehicledata.py:
```diff
@@ -37,7 +44,7 @@
     'retry_delay': timedelta(minutes=5),
     'retry_exponential_backoff': True, #Allow for progressive longer waits between retries
-    'on_failure_callback': partial(task_fail_slack_alert, owners = names),
+    'on_failure_callback': partial(task_fail_slack_alert, use_proxy = True),
```
Did you test this on morbius? Is it working?

tahaislam commented on volumes/vds/readme.md:
```diff
@@ -23,39 +24,106 @@
+Column `volume_veh_per_hr` stores the volumes in vehicles per hour for that bin. Note that different sensors have different bin width which affects the conversion from volume to count (see `vds.detector_expected_bins`). For details on converting `raw_vdsdata` to 15 minute counts, see `vds.counts_15min` or the corresponding insert script at `bdit_data-sources/volumes/vds/sql/insert/insert_counts_1 ...
 This table retains zero bins to enable potential future differente treatment of missing and zero values. 
-Contains `division_id IN (2, 8001)`. A one day sample for `division_id = 8001` was explored under the heading [vds.raw_vdsdata_div8001](#vdsraw_vdsdata_div8001); more investigation is needed to determine what this data can be used for. 
+Contains `division_id IN (2, 8001)`. A one day sample for `division_id = 8001` was explored under the heading [vds.raw_vdsdata_div8001](#vdsraw_vdsdata_div8001); more investigation is needed to determine what this data can be used for.  
```
not sure if this link `[vds.raw_vdsdata_div8001](#vdsraw_vdsdata_div8001)` is working as intended

tahaislam commented on volumes/vds/sql/select/select-itsc_vdsvehicledata_dst_safe.sql:
```diff
@@ -11,10 +11,9 @@ WHERE
+    ) OR (
+        timestamputc >= TIMEZONE('UTC', {start}::timestamptz + interval '3 HOURS') -- noqa: PRS
+        AND timestamputc < TIMEZONE('UTC', {start}::timestamptz + interval '1 DAY') -- noqa: PRS
+    ))
```
Do you skip this hour everyday or just that specific date in November?

tahaislam commented on volumes/vds/sql/tables/create-table-detector_inventory_cursed.sql:
```diff
@@ -0,0 +1,56 @@
+DROP TABLE vds.detector_inventory_cursed;
```
I think these `DROP` statements are dangerous unless you added it intentionally. Is it safe to keep this query in this file?

tahaislam reviewed (commented):
Looks great. I left a few comments and I think sqlfluff is complaining about one of the scripts as well.

radumas commented on volumes/vds/sql/views/create-view-detector_inventory.sql:
```diff
@@ -0,0 +1,91 @@
+DROP VIEW vds.detector_inventory;
```
I feel like some of these rules should be in look up tables instead, to make it a bit easier to be understood. 

Expected bins and RESCU detector codes translating to different areas for starters.

radumas commented on volumes/vds/readme.md:
```diff
@@ -0,0 +1,448 @@
+    - [Data Availability](#data-availability)
+    - [Future Work](#future-work)
+2. [Table Structure](#table-structure)
+    - [vds.raw_vdsdata](#vdsraw_vdsdata)
+    - [vds.counts_15min](#vdscounts_15min)
+    - [vds.counts_15min_bylane](#vdscounts_15min_bylane)
+    - [vds.raw_vdsvehicledata](#vdsraw_vdsvehicledata)
+    - [vds.veh_speeds_15min](#vdsveh_speeds_15min)
+    - [vds.veh_length_15min](#vdsveh_length_15min)
+    - [vds.vdsconfig](#vdsvdsconfig)
+    - [vds.entity_locations](#vdsentity_locations)
+    - [vds.detector_inventory](#vdsdetector_inventory)
+    - [vds.detector_inventory_cursed](#vdsdetector_inventory_cursed)
```
I think there are enough tables that sub-headings might be warranted. 

- Look up tables and views
- Aggregate Tables
- Raw Data
- Cursed

radumas commented on volumes/vds/readme.md:
```diff
@@ -0,0 +1,448 @@
+
+**update_inventories**  
+These tasks are arbitrarily grouped under this DAG because they run on the same schedule (daily) and with the same connections. They could instead be in a separate DAG if desired. 
+- `skip_update_inventories` task ensures these only run for the most recent schedule interval (doesn't backfill). 
```
ooh you gotta show us more of this in the presentation, probably other applications out there.

radumas reviewed (changes requested):
I focussed on the README since that's what's about to be presented.

I think there's a bit of tension between that presentation ("What's new with the RESCU pipeline?") vs. a README that will stand the test of time ("This is the VDS data set, here's how to use it"). 

I appreciate that the VDS README is... a lot, so we can leave aside incorporating it and purging the `rescu` folder until after.

My thought on general layout of these kinds of READMEs is (yes I should present this somewhere, along with my rules for an ideal pipeline, etc.).

2.     What is this?
3.     How do I access it? Where is the OpenData (if it exists)?
4.     How do I use it?
5.     How was it aggregated & filtered? What pitfalls should I avoid?
6.     Data Ops: how does the pipeline work?
7.     Data Ops: something went wrong predictably, how do I fix it?
8.     Data Ops: something went wrong that we can't fix, who should I contact?

For the table structure section, I think the tips for using section should come first, followed by the "static" tables, and then the data tables.

I think the summary of improvements section is great for the presentation, but probably not something to be referred to in the README in a few years. It could be put in to the PR Description and you could open with that in the presentation. 

Impressive work identifying patterns in the detector IDs!

I'll have more comments but my train is arriving.

tahaislam commented on dags/vds_pull_vdsdata.py:
```diff
@@ -135,8 +129,12 @@
+                external_task_id="starting_point"
+        )
+
+        [pull_detector_inventory_task, pull_entity_locations_task] >> t_done
```
This means the downstream DAG would be triggered when this task group succeeds not the whole DAG. Just confirming this's what you meant

tahaislam commented on dags/vds_pull_vdsvehicledata.py:
```diff
@@ -55,6 +56,15 @@
          template_searchpath=os.path.join(repo_path,'volumes/vds/sql'),
          schedule_interval='0 4 * * *') as dag: #daily at 4am
 
+    t_upstream_done = ExternalTaskSensor(
```
I think this DAG would usually fail to find the upstream DAG done and would wait for the second retry. If you want to avoid this behavior, you can add a delay between the two DAGs (and update the schedule of one of the two DAGs to compensate for this delay as well)

tahaislam commented on dags/vds_pull_vdsdata.py:
```diff
@@ -135,8 +129,12 @@
-    check_partitions
-    vdsdata >> v15data #pull then summarize
-    update_inventories
+    update_inventories >> check_partitions >> vdsdata >> v15data #pull then summarize
```
It might make more sense to have task group `update_inventories` defined at the beginning to reflect that it's the first task (group) in this DAG. But that doesn't affect anything in how the DAG works, i.e., you don't have to do what I'm suggesting

tahaislam reviewed (commented):
Looks great. just left three comments for you to check

radumas commented:
Just thinking we should update the READMEs for `rescu` at a basic level to indicate that pipeline isn't running anymore. Both at the root of this repo and at `volumes/rescu`

radumas commented on volumes/vds/readme.md:
```diff
@@ -0,0 +1,468 @@
+## How was it aggregated & filtered? What pitfalls should I avoid?
+- Detectors/time periods with no records will not appear in summary tables.
+- Some detectors have come and gone, for example the Allen. Check the sensor you want is available in the time period. 
+- Zeros/nulls/missing records are all considered zeros when aggregating to `counts_15min` and `counts_15min_bylane` in line with logic under previous `RESCU` pipeline. Aggregation method can be adjusted/examined by going back to `raw_vdsdata` table. 
```
Can add a reference here to this past work, until more automated checks and transformation are in place
https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/rescu/README.md#6--how-often-are-data-quality-assessment-processes-for-the-data-undertaken

gabrielwol commented:
> Just thinking we should update the READMEs for `rescu` at a basic level to indicate that pipeline isn't running anymore. Both at the root of this repo and at `volumes/rescu`

Added in this readme PR: https://github.com/CityofToronto/bdit_data-sources/pull/712

radumas reviewed (approved):
Woohoo! Looking good!
